src/Database/MagellanExpressions/MagellanBaseExpression.php

```diff
@@ -64,43 +84,93 @@ public function invoke($builder, string $bindingType, ?string $as = null): Expre
                 }
 
                 return new BindingExpression($param);
-            });
+            })->toArray();
 
-        return BuilderUtils::buildPostgisFunction($builder, $bindingType, $this->geometryType?->value, $this->postgisFunction, $as, ...$params);
-    }
+        $generatorClass = config('magellan.sql_generator', WKTGenerator::class);
+        $generator = new $generatorClass();
 
-    public function returnsGeometry(): bool
-    {
-        return $this instanceof MagellanGeometryExpression;
-    }
+        $geometryTypeCastAppend = $this->geometryType ? "::{$this->geometryType->value}" : '';
+        $preparedParameters = self::prepareParams($grammar, $generator, $geometryTypeCastAppend, $params);
+        $paramString = implode(', ', $preparedParameters);
 
-    public function returnsBbox(): bool
-    {
-        return $this instanceof MagellanBBoxExpression;
+        return Config::get('magellan.schema').'.'."$this->postgisFunction($paramString)";
     }
 
-    public function returnsSet(): bool
+    private function prepareParams(Grammar $grammar, BaseGenerator $generator, string $geometryTypeCastAppend, array $params): array
     {
-        return $this instanceof MagellanSetExpression;
-    }
+        return collect($params)->map(function ($param) use ($geometryTypeCastAppend, $generator, $grammar) {
+
+            // 1. Check if param is queryable -> it's a subquery
+            if ($this->isQueryable($param)) {
+                // --> Create sub and replace with param array
+                return $this->createSub($param);
+            }
+
+            // 2. Basic Binding Value
+            if ($param instanceof BindingExpression) {
+                // --> escape and replace
+                return $grammar->escape($param->getValue());
+            }
+
+            // 3. Geo Param
+            if ($param instanceof GeoParam) {
+                $value = $param->getValue();
+
+                if ($value instanceof MagellanBaseExpression) {
+                    return $value->getValue($grammar);
+                } elseif ($this->isQueryable($value)) {
+                    return $this->createSub($value).$geometryTypeCastAppend;
+                } elseif (is_array($value)) {
+                    $wrapped = array_map(fn ($geometry) => GeoParam::wrap($geometry), $value);
+                    $prepared = $this->prepareParams($grammar, $generator, $geometryTypeCastAppend, $wrapped);
+                    $imploded = implode(', ', $prepared);
+
+                    return "ARRAY[$imploded]";
+
+                } elseif ($value instanceof Geometry) {
+                    return $generator->toPostgisGeometrySql($value, Config::get('magellan.schema'));
```

Comment:
We need to append the type cast for all other cases too.

Reply:
As discussed yesterday we might add a new way of casting. 
Maybe we will not need the type cast appends here in generell.
Let's continue our discussion later